gqcp/include/Mathematical/Algorithm/IterativeAlgorithm.hpp

```diff
@@ -0,0 +1,96 @@
+// This file is part of GQCG-gqcp.
+// 
+// Copyright (C) 2017-2019  the GQCG developers
+// 
+// GQCG-gqcp is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+// 
+// GQCG-gqcp is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with GQCG-gqcp.  If not, see <http://www.gnu.org/licenses/>.
+// 
+#pragma once
+
+
+#include "Mathematical/Algorithm/IterationCycle.hpp"
+#include "Mathematical/Algorithm/ConvergenceCriterion.hpp"
+
+#include <cstddef>
+
+
+namespace GQCP {
+
+
+
+/**
+ *  An algorithm that performs iterations. In every iteration, convergence is checked and a set of iteration steps is performed.
+ * 
+ *  @param _Environment             the type of environment that this algorithm is associated to
+ */
+template <typename _Environment>
+class IterativeAlgorithm {
+public:
+    using Environment = _Environment;
+
+
+private:
+    size_t maximum_number_of_iterations;
+    size_t iteration = 0;  // the current iteration counter
+
+    IterationCycle<Environment> iteration_cycle;
```

Comment:
If I recall correctly, there was an idea to allow alternating cycles with different accelerators or dampeners instead of chaining them in one cycle. If you see fit you could already implement this, I have no idea of the advantageous of alternating cycles.

Reply:
If I understand it correctly, your alternating cycles could be implemented by effectively implementing the alternation in one bigger cycle.

One other approach could be to extend the current functionality by allowing for multiple alternating iteration cycles. I think it's a nice idea, but I won't implement this in this PR.